contracts/user/ConsumerWrapper.sol

```diff
@@ -0,0 +1,232 @@
+    
+    error InsufficientFund(uint256 fundAmount, uint256 requiredAmount);
+    error InvalidSubId();
+    error GasLimitTooBig(uint256 have, uint32 want);
+    error CallbackFailed(bytes32 requestId);
+
+    mapping(address => uint64) public userSubIds;
+    mapping(bytes32 => RequestData) public pendingRequests;
+    
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address adapter) BasicRandcastConsumerBase(adapter) {
+        _disableInitializers();
+    }
+
+    function initialize() public initializer {
+        __Ownable_init();
+    }
+    // solhint-disable-next-line no-empty-blocks
+
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    function estimateFee(uint64 subId, uint32 callbackGasLimit)
+        public
+        view
+        returns (uint256 requestFee)
+    {
+        uint32 overhead = _FEE_OVERHEAD;
+        if (subId == 0) {
+            subId = userSubIds[msg.sender];
+            if (subId == 0) {
+                return IAdapter(adapter).estimatePaymentAmountInETH(
+                    callbackGasLimit, overhead, 0, tx.gasprice * 3, _GROUP_SIZE
+                );
+            }
+        }
+        // Get subscription details only if subId is not zero
+        Subscription memory sub = _getSubscription(subId);
+        uint32 tierFee;
+        if (sub.freeRequestCount == 0) {
+            tierFee = _calculateTierFee(sub.reqCount, sub.lastRequestTimestamp, sub.reqCountInCurrentPeriod);
+        }
+        uint256 estimatedFee = IAdapter(adapter).estimatePaymentAmountInETH(
+            callbackGasLimit, overhead, tierFee, tx.gasprice * 3, _GROUP_SIZE
+        );
+        return estimatedFee > (sub.balance - sub.inflightCost) ? estimatedFee - (sub.balance - sub.inflightCost) : 0;
+    }
+
+    function cancelSubscription() external {
+        uint64 subId = userSubIds[msg.sender];
+        if (subId == 0) {
+            return;
+        }
+        IAdapter(adapter).cancelSubscription(subId, msg.sender);
+        delete userSubIds[msg.sender];
+    }
+
+    function _fundSubId(uint64 subId, uint32 callbackGasLimit) internal returns (uint64) {
+
+        if (subId == 0) {
+            subId = userSubIds[msg.sender];
+            if (subId == 0) {
+                subId = IAdapter(adapter).createSubscription();
+                IAdapter(adapter).addConsumer(subId, address(this));
+                userSubIds[msg.sender] = subId;
+            }
+        }
+        uint256 requiredAmount = estimateFee(subId, callbackGasLimit);
+        if (msg.value < requiredAmount) {
+            revert InsufficientFund({fundAmount: msg.value, requiredAmount: requiredAmount});
+        }
+        if (requiredAmount == 0) {
+            return subId;
+        }
+        IAdapter(adapter).fundSubscription{value: msg.value}(subId);
+        return subId;
+    }
+
+    function _getSubscription(uint64 subId) internal view returns (Subscription memory sub) {
```

Comment:
Let's add an external view function, to allow users query their subscription with their address instead of subId managed internally by the wrapper.

like `function getSubscription(address user) external view returns (Subscription memory sub)`